File: LeetCode root folder 6/Maximize Palindrome Length From Subsequences.py

# don't know why it's so slow: MLE
class Solution:
    def longestPalindrome(self, word1: str, word2: str) -> int:
        a = word1 + word2
        n = len(a)
        dp = [[[[-1]*2 for _ in range(2)] for _ in range(n)] for _ in range(n)]
        def recursive(i, j, f1, f2):
            if dp[i][j][f1][f2] != -1: return dp[i][j][f1][f2]
            if i == j: return 1 if f1 and f2 else -inf
            if i > j: return 0 if f1 and f2 else -inf
            if a[i] == a[j]:
                if i in range(len(word1)): f1 = 1
                if j in range(len(word1), n): f2 = 1
                res = recursive(i+1, j-1, f1, f2) + 2
            else:
                res = max(
                    recursive(i+1, j, f1, f2),
                    recursive(i, j-1, f1, f2)
                )
            dp[i][j][f1][f2] = res
            return res
        res = recursive(0, n-1, 0, 0)
        return res if res != -inf else 0

File: LeetCode root folder 6/Maximize Palindrome Length From Subsequences.cpp
// I just cannot believe how fking faster it is to use static array dp over dynamic size vector even if it means to
// create dp array of size 16 millions for every single fucking testcase...
// this is so fking dramatic: 12%

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int longestPalindrome(string word1, string word2) {
        string a = word1 + word2;
        int n = a.size();
        const int NEG_INF = -1000000000;

        // dp[i][j][f1][f2]
        static int dp[2000][2000][2][2];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                for (int f1 = 0; f1 < 2; f1++)
                    for (int f2 = 0; f2 < 2; f2++)
                        dp[i][j][f1][f2] = -1;

        function<int(int,int,int,int)> recursive = [&](int i, int j, int f1, int f2) -> int {
            if (dp[i][j][f1][f2] != -1) return dp[i][j][f1][f2];
            if (i == j) return dp[i][j][f1][f2] = (f1 && f2) ? 1 : NEG_INF;
            if (i > j) return dp[i][j][f1][f2] = (f1 && f2) ? 0 : NEG_INF;

            int res;
            if (a[i] == a[j]) {
                int nf1 = f1, nf2 = f2;
                if (i < (int)word1.size()) nf1 = 1;
                if (j >= (int)word1.size()) nf2 = 1;
                res = recursive(i + 1, j - 1, nf1, nf2) + 2;
            } else {
                res = max(recursive(i + 1, j, f1, f2),
                          recursive(i, j - 1, f1, f2));
            }
            return dp[i][j][f1][f2] = res;
        };

        int res = recursive(0, n - 1, 0, 0);
        return (res > 0) ? res : 0;
    }
};